File: sparkSql/join/UkMakerSpaces.scala
package com.sparkTutorial.sparkSql.join

import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.{SparkSession, functions}

object UkMakerSpaces {

  def main(args: Array[String]) {

    Logger.getLogger("org").setLevel(Level.ERROR)

    val session = SparkSession.builder().appName("UkMakerSpaces").master("local[*]").getOrCreate()

    val makerSpace = session.read.option("header", "true").csv("in/uk-makerspaces-identifiable-data.csv")

    val postCode = session.read.option("header", "true").csv("in/uk-postcode.csv")
       .withColumn("PostCode", functions.concat_ws("", functions.col("PostCode"), functions.lit(" ")))

    System.out.println("=== Print 20 records of makerspace table ===")
    makerSpace.select("Name of makerspace", "Postcode").show()

    System.out.println("=== Print 20 records of postcode table ===")
    postCode.show()

    val joined = makerSpace.join(postCode, makerSpace.col("Postcode").startsWith(postCode.col("Postcode")), "left_outer")

    System.out.println("=== Group by Region ===")
    joined.groupBy("Region").count().show(200)
  }
}

File: sparkSql/join/UkMakerSpaces.py
from pyspark.sql import SparkSession, functions as fs

if __name__ == "__main__":
    session = SparkSession.builder.appName("UkMakerSpaces").master("local[*]").getOrCreate()

    makerSpace = session.read.option("header", "true") \
        .csv("in/uk-makerspaces-identifiable-data.csv")

    postCode = session.read.option("header", "true").csv("in/uk-postcode.csv") \
        .withColumn("PostCode", fs.concat_ws("", fs.col("PostCode"), fs.lit(" ")))

    print("=== Print 20 records of makerspace table ===")
    makerSpace.select("Name of makerspace", "Postcode").show()

    print("=== Print 20 records of postcode table ===")
    postCode.select("PostCode", "Region").show()

    joined = makerSpace \
        .join(postCode, makerSpace["Postcode"].startswith(postCode["Postcode"]), "left_outer")

    print("=== Group by Region ===")
    joined.groupBy("Region").count().show(200)